operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+		if slices.IndexFunc(sv.Status.LoadBalancer.Ingress, func(in core_v1.LoadBalancerIngress) bool {
+			return net.ParseIP(in.IP).Equal(alloc.IP)
+		}) == -1 {
+			sv.Status.LoadBalancer.Ingress = append(sv.Status.LoadBalancer.Ingress, core_v1.LoadBalancerIngress{
+				IP: alloc.IP.String(),
+			})
+			statusModified = true
+		}
+	}
+
+	if sv.isSatisfied() {
+		if ipam.setSVCSatisfiedCondition(sv, true, "satisfied", "") {
+			statusModified = true
+		}
+	}
+
+	ipam.serviceStore.Upsert(sv)
+
+	return statusModified, err
+}
+
+func (ipam *LBIPAM) setSVCSatisfiedCondition(
+	sv *ServiceView,
+	satisfied bool,
+	reason, message string,
+) (statusModified bool) {
+	status := meta_v1.ConditionFalse
+	if satisfied {
+		status = meta_v1.ConditionTrue
+	}
+
+	for _, cond := range sv.Status.Conditions {
+		if cond.Type == ciliumSvcRequestSatisfiedCondition &&
+			cond.Status == status &&
+			cond.ObservedGeneration == sv.Generation &&
+			cond.Reason == reason &&
+			cond.Message == message {
+			return false
+		}
+	}
+
+	sv.Status.Conditions = append(sv.Status.Conditions, meta_v1.Condition{
+		Type:               ciliumSvcRequestSatisfiedCondition,
+		Status:             status,
+		ObservedGeneration: sv.Generation,
+		LastTransitionTime: meta_v1.Now(),
+		Reason:             reason,
+		Message:            message,
+	})
+	return true
+}
+
+func (ipam *LBIPAM) findRangeOfIP(sv *ServiceView, ip net.IP) (lbRange *LBRange, foundPool bool, err error) {
+	for _, r := range ipam.rangesStore.ranges {
+		if r.Disabled() {
+			continue
+		}
+
+		cidr := r.allocRange.CIDR()
+		if !cidr.Contains(ip) {
+			continue
+		}
+
+		pool, found := ipam.poolStore.GetByName(r.originPool)
+		if !found {
+			continue
+		}
+
+		foundPool = true
+
+		if pool.Spec.ServiceSelector != nil {
+			selector, err := slim_meta_v1.LabelSelectorAsSelector(pool.Spec.ServiceSelector)
+			if err != nil {
+				return nil, false, fmt.Errorf("Making selector from pool '%s' label selector: %w", pool.Name, err)
+			}
+
+			if !selector.Matches(sv.Labels) {
+				continue
+			}
+		}
+
+		return r, false, nil
+	}
+
+	return nil, false, nil
+}
+
+// isResponsibleForSVC checks if LB IPAM should allocate and assign IPs or some other controller
+func (ipam *LBIPAM) isResponsibleForSVC(svc *core_v1.Service) bool {
+	// Ignore non-lb services
+	if svc.Spec.Type != core_v1.ServiceTypeLoadBalancer {
+		return false
+	}
+
+	// We will assume that we are the default LB, LB-IPAM shouldn't be enabled clusters that don't support LBClasses
+	// and have multiple LBs.
+	if svc.Spec.LoadBalancerClass == nil {
+		return true
+	}
+
+	if !slices.Contains(ipam.lbClasses, *svc.Spec.LoadBalancerClass) {
+		return false
+	}
+
+	return true
+}
+
+type AddressFamily string
+
+const (
+	IPv4Family AddressFamily = "IPv4"
+	IPv6Family AddressFamily = "IPv6"
+)
+
+func (ipam *LBIPAM) allocateIPAddress(
+	sv *ServiceView,
+	family AddressFamily,
+) (
+	newIP *net.IP,
+	chosenRange *LBRange,
+	full bool,
```

Comment:
Why not just return a specific constant Error for the "no pool" and "all pools full" state?